UIforETW/UIforETWDlg.cpp

```diff
@@ -40,10 +41,821 @@ const int kRecordTraceHotKey = 1234;
+	std::wstring result = default;
+
+	if (!rawDirectory.empty())
+		result = rawDirectory;
+
+	// Make sure the name ends with a backslash.
+	if ((!result.empty()) && (result[ result.size( ) - 1 ] != '\\'))
+		result += '\\';
+	return result;
+}
+
+
+bool enhancedCreateDirectory( _In_ std::wstring pathName )
+{
+	//(first parameter):
+	//    There is a default string size limit for paths of 248 characters.
+	//    To extend this limit to 32,767 wide characters,
+	//        call the Unicode version of the function
+	//        and prepend "\\?\" to the path.
+	//                    ^ ==> L"\\\\?\\"
+
+	//TODO: what if network path?
+	//should we check if just starts with L"\\\\"?
+	if (pathName.compare( 0, 4, L"\\\\?\\" ) != 0)
+		pathName = (L"\\\\?\\" + pathName);
+
+	const std::wstring& longPath = pathName;
+
+	//If [CreateDirectoryW] succeeds, the return value is nonzero.
+	//If [CreateDirectoryW] fails, the return value is zero.
+	//To get extended error information, call GetLastError.
+	//Possible errors include the following:
+	//    ERROR_ALREADY_EXISTS
+	//    ERROR_PATH_NOT_FOUND
+	const BOOL directoryCreated = CreateDirectoryW(longPath.c_str( ), NULL);
+
+	if (directoryCreated != 0)
+		return true;
+	const DWORD lastErr = GetLastError();
+	if (lastErr == ERROR_ALREADY_EXISTS)
+		return false;
+	if (lastErr == ERROR_PATH_NOT_FOUND)
+		return false;
+
+	handleUnexpectedCreateDirectory(longPath, lastErr);
+	return false;
+}
+
+bool isValidDirectory( _In_ std::wstring path )
+{
+	//PathIsDirectoryW expects a path thats no longer than MAX_PATH
+	if (path.size( ) < MAX_PATH)
+	{
+		//[PathIsDirectory] returns (BOOL)FILE_ATTRIBUTE_DIRECTORY if the path 
+		//is a valid directory; otherwise, FALSE.
+		const BOOL pathIsDir = PathIsDirectoryW(path.c_str());
+		if (pathIsDir == ( (BOOL)( FILE_ATTRIBUTE_DIRECTORY ) ))
+		{
+			return true;
+		}
+		return false;
+	}
+	//TODO: what if network path?
+	//should we check if just starts with L"\\\\"?
+	if (path.compare( 0, 4, L"\\\\?\\" ) != 0)
+		path = (L"\\\\?\\" + path);
+
+	const std::wstring& longPath = path;
+	const bool isPathAtAllValid = isValidPathToFSObject(longPath);
+
+	if (!isPathAtAllValid)
+		return false;
+	//If [GetFileAttributes] fails, the return value is INVALID_FILE_ATTRIBUTES.
+	//To get extended error information, call GetLastError.
+	const DWORD pathAttributes = GetFileAttributesW(longPath.c_str());
+	if (pathAttributes == INVALID_FILE_ATTRIBUTES)
+	{
+		const DWORD lastErr = GetLastError();
+		if (lastErr == ERROR_BAD_NETPATH)
+		{
+			outputPrintf(L"Oops! %s is NOT a valid directory, "
+						 L"it's a network share! GetFileAttributes "
+						 L"needs a subfolder of it!\n", longPath.c_str( ));
+		}
+		return false;
+	}
+	if ((pathAttributes & FILE_ATTRIBUTE_DIRECTORY) == 0)
+		return false;
+	//all tests passed!
+	return true;
+}
+
+std::wstring getKernelStackWalk( _In_ const bool bSampledStacks, _In_ const bool bCswitchStacks )

```

Comment:
I understand the desire to not have a monster function that sets all the profile settings, but is this really the best way to break it up? It seems like it actually reduces clarity by spreading the code out too much.